text/0813-template-layering.md

```diff
@@ -0,0 +1,332 @@
+
+   ```
+   import Foo form 'somewhere';
+
+   export const Bar = <template><Foo /><template>;
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export const Bar = template(`<Foo />`, () => ({ Foo }));
+   ```
+
+3. Class
+
+   ```
+   import Foo form 'somewhere';
+
+   export default class Bar {
+     <template><Foo /><template>
+   }
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export default @template(`<Foo />`, () => ({ Foo }) class Bar {
+   }
+   ```
+
+In these small snippets, the scope bindings may look very verbose compared to
+the size of the templates, but in real-world templates, the ratio will improve.
+
+**TBD**: I think we should amend the v2 addon spec to make this the recommended
+publishing format.
+
+**Note**: Whether we actually implement `<template>` using this desugaring is
+an internal implementation detail. However, the semantics should work and users
+should be able to reason about `<template>` with this desugaring in mind.
+
+### Tagged Template Literals
+
+1. Top-level declaration
+
+   ```
+   import Foo form 'somewhere';
+
+   <template>
+     <Foo />
+   <template>
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export default template`<Foo />`;
+   ```
+
+2. Expression
+
+   ```
+   import Foo form 'somewhere';
+
+   export const Bar = template`<Foo />`;
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export const Bar = template`<Foo />`;
+   ```
+
+3. Class
+
+   ```
+   import Foo form 'somewhere';
+
+   export default class Bar {
+     <template><Foo /><template>
+   }
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export default class Bar {
+     static { template`<Foo />` }
+   }
+   ```
+
+**TBD**
+
+## How we teach this
+
+**TBD**: While the `template()` function is a bit cumbersome to use in practice
+I think what it does is in fact quite easy to understand and teach. It may
+actually help with teaching `<template>` if we can show what it desugars into.
+
+## Drawbacks
+
+**TBD**: Having the ability to supply the variable binding closure manually may
+open it up to misuse (i.e. the closure does more than just simply capturing the
+surrounding lexical scope, or does so in ways we didn't intend). However, this
+is not a new problem in a sense, as the primitives already exists, and the
+status quo is "if you cannot use `<template>`, use the primitives directly.
+Also, with default helper manager, you can pretty much do anything along those
+lines anyway with an inline closure that you invoke inside the template.
+
+## Alternatives
+
+**TBD**
+
+## Unresolved questions
+
+**TBD**
```

Comment:
would we have the ability to use this at runtime?

I imagine this `template` function _always_ converts to the underlying format, but in apps, a build step could swap out the implementation -- allowing `template` to exist at runtime, if imported and used dynamically, such as the use cases of:
 - https://github.com/NullVoxPopuli/ember-repl
 - [limber.glimdown.com/](https://limber.glimdown.com/)

Reply:
Yes that's the idea, see:

> It is designed to be pre-processed at build time where possible, just like the `<template>` feature and today's "official" `hbs` tag. However, in cases where this is not possible, there will be a runtime implementation available provided the template compiler is also available at runtime.

Reply:
Which is why it's important that the `template()` version doesn't do any magic that is not actually possible at runtime.

Reply:
YAAAS